[Site-Isolation] RemoteDOMWindow::setLocation should have identical security checks as LocalDOMWindow::setLocation

nipunshukla21 opened the pull request:
#### b3ce2e746ba582e3fde8554c1c97bd01914d2b76
<pre>
[Site-Isolation] RemoteDOMWindow::setLocation should have identical security checks as LocalDOMWindow::setLocation
<a href="https://rdar.apple.com/116500603">rdar://116500603</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=296457">https://bugs.webkit.org/show_bug.cgi?id=296457</a>

Reviewed by Alex Christensen.

Move a number of functions which check security of setLocation from LocalDOMWindow to DOMWindow parent class and invoke them in a new setLocationSecurityChecks function which is called in both LocalDOMWindow::setLocation() and RemoteDOMWindow::setLocation().

* LayoutTests/http/tests/site-isolation/remotedomwindow-setlocation-passes-security-checks-expected.txt: Added.
* LayoutTests/http/tests/site-isolation/remotedomwindow-setlocation-passes-security-checks.html: Added.
* Source/WebCore/page/DOMWindow.cpp:
(WebCore::DOMWindow::protectedDocumentIfLocal):
(WebCore::DOMWindow::isCurrentlyDisplayedInFrame const):
(WebCore::DOMWindow::printErrorMessage const):
(WebCore::DOMWindow::crossDomainAccessErrorMessage):
(WebCore::DOMWindow::isInsecureScriptAccess):
(WebCore::DOMWindow::passesSetLocationSecurityChecks):
* Source/WebCore/page/DOMWindow.h:
* Source/WebCore/page/LocalDOMWindow.cpp:
(WebCore::LocalDOMWindow::find const):
(WebCore::didAddStorageEventListener):
(WebCore::LocalDOMWindow::setLocation):
(WebCore::LocalDOMWindow::isCurrentlyDisplayedInFrame const): Deleted.
(WebCore::LocalDOMWindow::printErrorMessage const): Deleted.
(WebCore::LocalDOMWindow::crossDomainAccessErrorMessage): Deleted.
(WebCore::LocalDOMWindow::isInsecureScriptAccess): Deleted.
* Source/WebCore/page/LocalDOMWindow.h:
* Source/WebCore/page/RemoteDOMWindow.cpp:
(WebCore::RemoteDOMWindow::setLocation):
* Source/WebCore/page/RemoteFrame.cpp:
(WebCore::RemoteFrame::frameDocumentSecurityOriginOrOpaque const):
* Source/WebCore/page/RemoteFrame.h:

Canonical link: <a href="https://commits.webkit.org/298598@main">https://commits.webkit.org/298598@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0f5e5f9e8e8b52cade971362a9717682a19be4cb

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/114527 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/34273 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/24736 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/120691 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/65251 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/116416 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/34902 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/42833 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/87044 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/41952 "") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/117475 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/27810 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/102852 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/67435 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/26995 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/20979 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/64373 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/97187 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/21093 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/123898 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/41540 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/31003 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/95862 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/41917 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/99043 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/95646 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/40812 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/18649 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/37583 "The change is no longer eligible for processing.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/18538 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/41419 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/46929 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/41005 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/44312 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/42755 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebCore/page/DOMWindow.cpp:
```diff
@@ -922,4 +922,124 @@ ExceptionOr<PushManager&> DOMWindow::pushManager()
+    RefPtr<Document> document;
+    if (this->document().hasException())
+        return String();
+    document = this->document().returnValue();
```
We are calling this->document() unnecessarily twice.

achristensen07 commented on Source/WebCore/page/DOMWindow.cpp:
```diff
@@ -922,4 +922,124 @@ ExceptionOr<PushManager&> DOMWindow::pushManager()
+    return true;
+}
+
+bool DOMWindow::setLocationSecurityChecks(LocalDOMWindow& activeWindow, const URL& completedURL, CanNavigateState& navigationState)
```
Can this be called passesSetLocationSecurityChecks or something that indicates that it is not setting the location but checking security?

achristensen07 commented on Source/WebCore/page/DOMWindow.cpp:
```diff
@@ -922,4 +922,124 @@ ExceptionOr<PushManager&> DOMWindow::pushManager()
+        return String();
+
+    RefPtr<Document> document;
+    if (this->document().hasException())
```
This always returns an exception if the DOMWindow is a RemoteDOMWindow.  I'm not sure that's what we want.

achristensen07 commented on Source/WebCore/page/DOMWindow.cpp:
```diff
@@ -922,4 +922,124 @@ ExceptionOr<PushManager&> DOMWindow::pushManager()
+        RefPtr<Document> document;
+        if (this->document().hasException())
+            return true;
+        document = this->document().returnValue();
```
Again, calling this->document() unnecessarily twice, always returns an exception for RemoteDOMWindow.

achristensen07 commented on Source/WebCore/page/DOMWindow.h:
```diff
@@ -223,7 +224,17 @@ class DOMWindow : public RefCounted<DOMWindow>, public EventTarget {
+    String crossDomainAccessErrorMessage(const LocalDOMWindow& activeWindow, IncludeTargetOrigin);
+
 protected:
+    bool isInsecureScriptAccess(LocalDOMWindow& activeWindow, const String& urlString);
```
Can we make activeWindow a const LocalDOMWindow& instead?  Same with setLocationSecurityChecks and navigationState below.

nipunshukla21 replied:
Done with the exception of navigationState, as we do want to modify that on occasion.

achristensen07 reviewed (commented):
Are there any changes in behavior for RemoteDOMWindows?  Can this include a test that failed before this change but passes after this change?

nipunshukla21 commented on Source/WebCore/page/DOMWindow.cpp:
```diff
@@ -922,4 +930,125 @@ ExceptionOr<PushManager&> DOMWindow::pushManager()
+    // Sandbox errors: Use the origin of the frames' location, rather than their actual origin (since we know that at least one will be "null").
+    URL activeURL = activeWindow.document()->url();
+    RefPtr<const SecurityOrigin> remoteFrameSecurityOrigin = (m_type == DOMWindowType::Remote) ? remoteFrame->frameDocumentSecurityOriginRef() : RefPtr<const SecurityOrigin>();
+    URL targetURL = localDocument ? localDocument->url() : remoteFrameSecurityOrigin->toURL();
```
This is kind of messy but I had the goal of not breaking encapsulation of the security origin so I made a function which returns a constant reference. Then that led to a Safer CPP violation -- making it necessary to do this. Is there a cleaner way?

achristensen07 replied:
This URL is only used to access the protocol and to create a SecurityOrigin.  Why not just leave it as a SecurityOrigin instead of going to a URL and back?

nipunshukla21 replied:
Unfortunately it looks like this has to be turned into a URL and then used to create a SecurityOrigin again, otherwise there are test failures regarding sandboxing as doing `SecurityOrigin.toString()` returns null.

achristensen07 replied:
Oh, I see.  The URL is indeed used for more than that.

achristensen07 commented on Source/WebCore/page/RemoteFrame.cpp:
```diff
@@ -35,6 +35,7 @@
 #include "RemoteDOMWindow.h"
 #include "RemoteFrameClient.h"
 #include "RemoteFrameView.h"
+#include "page/SecurityOrigin.h"
```
This shouldn't have page/ in the include path.

achristensen07 commented on Source/WebCore/page/RemoteFrame.cpp:
```diff
@@ -176,6 +177,14 @@ RefPtr<SecurityOrigin> RemoteFrame::frameDocumentSecurityOrigin() const
     return frameTreeSyncData().frameDocumentSecurityOrigin;
 }
 
+const SecurityOrigin& RemoteFrame::frameDocumentSecurityOriginRef() const
```
frameDocumentSecurityOriginOrOpaque

achristensen07 commented on Source/WebCore/page/DOMWindow.cpp:
```diff
@@ -922,4 +930,125 @@ ExceptionOr<PushManager&> DOMWindow::pushManager()
+        // Can we name the SecurityOrigin function better to make this more clear?
+
+        // This check only makes sense with LocalDOMWindows as RemoteDOMWindows necessarily have different origins
+        if (m_type == DOMWindowType::Local && activeWindow.protectedDocument()->protectedSecurityOrigin()->isSameOriginDomain(protectedDocumentIfLocal()->protectedSecurityOrigin()))
```
May as well null check this document like we do in DOMWindow::passesSetLocationSecurityChecks

achristensen07 commented on Source/WebCore/page/DOMWindow.cpp:
```diff
@@ -922,4 +930,126 @@ ExceptionOr<PushManager&> DOMWindow::pushManager()
+bool DOMWindow::passesSetLocationSecurityChecks(const LocalDOMWindow& activeWindow, const URL& completedURL, CanNavigateState& navigationState)
+{
+    ASSERT(navigationState != CanNavigateState::Unchecked);
+    if (!isCurrentlyDisplayedInFrame())
```
Ok, so going through this, this is the only check that was missing from the Remote version, but if the frame is a RemoteFrame we know its canNavigate will fail so it's redundant.  This is a good change anyways because it takes shared logic and shares it rather than duplicates it, so future developers won't forget one of the copies.

nipunshukla21 committed:
[Site-Isolation] RemoteDOMWindow::setLocation should have identical security checks as LocalDOMWindow::setLocation
rdar://116500603
https://bugs.webkit.org/show_bug.cgi?id=296457

Reviewed by Alex Christensen.

Move a number of functions which check security of setLocation from LocalDOMWindow to DOMWindow parent class and invoke them in a new setLocationSecurityChecks function which is called in both LocalDOMWindow::setLocation() and RemoteDOMWindow::setLocation().

* LayoutTests/http/tests/site-isolation/remotedomwindow-setlocation-passes-security-checks-expected.txt: Added.
* LayoutTests/http/tests/site-isolation/remotedomwindow-setlocation-passes-security-checks.html: Added.
* Source/WebCore/page/DOMWindow.cpp:
(WebCore::DOMWindow::protectedDocumentIfLocal):
(WebCore::DOMWindow::isCurrentlyDisplayedInFrame const):
(WebCore::DOMWindow::printErrorMessage const):
(WebCore::DOMWindow::crossDomainAccessErrorMessage):
(WebCore::DOMWindow::isInsecureScriptAccess):
(WebCore::DOMWindow::passesSetLocationSecurityChecks):
* Source/WebCore/page/DOMWindow.h:
* Source/WebCore/page/LocalDOMWindow.cpp:
(WebCore::LocalDOMWindow::find const):
(WebCore::didAddStorageEventListener):
(WebCore::LocalDOMWindow::setLocation):
(WebCore::LocalDOMWindow::isCurrentlyDisplayedInFrame const): Deleted.
(WebCore::LocalDOMWindow::printErrorMessage const): Deleted.
(WebCore::LocalDOMWindow::crossDomainAccessErrorMessage): Deleted.
(WebCore::LocalDOMWindow::isInsecureScriptAccess): Deleted.
* Source/WebCore/page/LocalDOMWindow.h:
* Source/WebCore/page/RemoteDOMWindow.cpp:
(WebCore::RemoteDOMWindow::setLocation):
* Source/WebCore/page/RemoteFrame.cpp:
(WebCore::RemoteFrame::frameDocumentSecurityOriginOrOpaque const):
* Source/WebCore/page/RemoteFrame.h:

Canonical link: https://commits.webkit.org/298598@main
